hir/src/asm.rs

```diff
@@ -0,0 +1,2447 @@
+
+    /// Pushes an element representing an unsigned 32-bit integer on the stack
+    pub fn push_u32(self, imm: u32) {
+        self.stack.push(Type::U32);
+        self.asm.push(self.ip, MasmOp::PushU32(imm));
+    }
+
+    /// Drops the element on the top of the stack
+    pub fn drop(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::Drop);
+    }
+
+    /// Drops the word (first four elements) on the top of the stack
+    pub fn dropw(self) {
+        self.stack.dropw();
+        self.asm.push(self.ip, MasmOp::Dropw);
+    }
+
+    /// Duplicates the `n`th element from the top of the stack, to the top of the stack
+    ///
+    /// A `n` of zero, duplicates the element on top of the stack
+    ///
+    /// The valid range for `n` is 0..=15
+    pub fn dup(self, n: usize) {
+        self.stack.dup(n);
+        self.asm.push(self.ip, MasmOp::Dup(n as u8));
+    }
+
+    /// Duplicates the `n`th word from the top of the stack, to the top of the stack
+    ///
+    /// A `n` of zero, duplicates the word on top of the stack
+    ///
+    /// The valid range for `n` is 0..=3
+    pub fn dupw(self, n: usize) {
+        self.stack.dupw(n);
+        self.asm.push(self.ip, MasmOp::Dupw(n as u8));
+    }
+
+    /// Swaps the `n`th element and the element on top of the stack
+    ///
+    /// The valid range for `n` is 1..=15
+    pub fn swap(self, n: usize) {
+        self.stack.swap(n);
+        self.asm.push(self.ip, MasmOp::Swap(n as u8));
+    }
+
+    /// Swaps the `n`th word and the word on top of the stack
+    ///
+    /// The valid range for `n` is 1..=3
+    pub fn swapw(self, n: usize) {
+        self.stack.swapw(n);
+        self.asm.push(self.ip, MasmOp::Swapw(n as u8));
+    }
+
+    /// Moves the `n`th element to the top of the stack
+    ///
+    /// The valid range for `n` is 2..=15
+    pub fn movup(self, idx: usize) {
+        self.stack.movup(idx);
+        self.asm.push(self.ip, MasmOp::Movup(idx as u8));
+    }
+
+    /// Moves the `n`th word to the top of the stack
+    ///
+    /// The valid range for `n` is 2..=3
+    pub fn movupw(self, idx: usize) {
+        self.stack.movupw(idx);
+        self.asm.push(self.ip, MasmOp::Movupw(idx as u8));
+    }
+
+    /// Moves the element on top of the stack, making it the `n`th element
+    ///
+    /// The valid range for `n` is 2..=15
+    pub fn movdn(self, idx: usize) {
+        self.stack.movdn(idx);
+        self.asm.push(self.ip, MasmOp::Movdn(idx as u8));
+    }
+
+    /// Moves the word on top of the stack, making it the `n`th word
+    ///
+    /// The valid range for `n` is 2..=3
+    pub fn movdnw(self, idx: usize) {
+        self.stack.movdnw(idx);
+        self.asm.push(self.ip, MasmOp::Movdnw(idx as u8));
+    }
+
+    /// Pops a boolean element off the stack, and swaps the top two elements
+    /// on the stack if that boolean is true.
+    ///
+    /// Traps if the conditional is not 0 or 1.
+    pub fn cswap(self) {
+        self.stack.drop();
```

Comment:
I get that this keeps track of the stack height, but if the stack elements are `Type` objects, then how do you keep track of the elements when the swap is conditional?

AFAICT there is no way to keep track of that, so it should be made clear exactly what information we are keeping track of (is it just the stack height? If so, why are the stack elements of type `Type`?).

The same question applies to all the conditional ops.

Reply:
What is documented here (and the default implementation) is what happens to the operand stack in the VM when that operation is performed. When using a type stack, we're tracking that behavior in the abstract, and we are only primarily concerned with the stack height. I should note that we are only ever pushing `Type::Felt` in this case, so it doesn't actually matter what the conditional or the operands are, as it doesn't change the semantics in the slightest.

If we were to expand the use of type stacks to something more general purpose, then we'd likely override the implementation of certain functions in the trait, to tailor them to the semantics of a type checker (i.e. validating that the operands of an operation like `cswap` are the same). The default implementation simply performs the necessary stack maintenance.